app/controllers/companies_controller.rb

```diff
@@ -13,8 +13,35 @@ def show
 
   def update_companies_list
     companies = GPW::Client.new.fetch_companies_list
+    companies.compact!
     companies.each { |company| Company.find_or_create_by!(company) }
+
     flash[:notice] = 'Companies list successfully updated'
     redirect_to companies_url
   end
+
+  def update_companies_details
+    companies = GPW::Client.new.fetch_gpw_companies_details
+    companies.compact!
+    companies.each do |company|
+      Company.find_by(isin: company[:isin])&.update(
+        stock: company[:stock],
+        pbv: company[:pbv],
+        pe: company[:pe],
+        condition: company[:condition],
+      )
+    end
+    companies = GPW::Client.new.fetch_nc_companies_details
+    companies.compact!
+    companies.each do |company|
+      Company.find_by(isin: company[:isin])&.update(
+        full_name: company[:full_name],
+        abbreviation: company[:abbreviation],
+        stock: company[:stock],
+        condition: company[:condition],
+      )
+    end
```

Comment:
Since the logic here is getting bigger, you can move it to a separate class. The class that does an operation (such as "updating companies") is called a "service" (you can read about service objects in Ruby).

A common practice is to have a service that handles a single operation (to keep the class small and to put a single well-defined responsibility into a single class). Then you just need a single public method in such class and people usually call such method `call`. In the other words, the class has a meaningful name and you do not need specific method name.

```ruby
CompaniesDetailsUpdator.new.call
```

The definition of the service class should be placed in the `lib` directory.

Reply:
that's fine probably. I don't like to call the directory `services` because I put there more things that I wouldn't call services (such as API clients, form objects, value objects, etc.). Personally I would go with `app/lib` if you want to have it in the `app` :)

But I'm aware some people put it into `app/services` and then I often see `app/form_objects`, etc. in their projects. They end up with one functionality divided into 10 directories. I prefer single `lib` directory for all the code responsible for non-web part of my project and then divide things there based on different parts of the apps (such as "authorization", "companies", "blog", etc.) instead of types of objects ("services", "form objects", etc.). I think we could discuss it more on voice call one day.

Reply:
I think your way is better so I will put this files to the `lib` directory, anyway it will be good if we discuss it on call